File: clinic-app/src/domains/maternity/services/maternityService.ts
import client from '@/api/client';
import { PurposeOfUse, VisitStatus } from '@/shared/enums';
import { VisitResponse } from '@/shared/types';

export interface MaternityDeliveryUpsert {
  action?: 'SAVE_DRAFT' | 'SIGN';
  episode_id?: string | null;
  delivered_at?: string | null;
  mode_of_delivery?: 'SVD' | 'C_SECTION' | 'ASSISTED' | 'UNKNOWN';
  outcome?: 'LIVE_BIRTH' | 'STILLBIRTH' | 'NEONATAL_DEATH' | 'UNKNOWN';
  baby_sex?: 'MALE' | 'FEMALE' | 'UNKNOWN';
  baby_weight_kg?: number | null;
  apgar_1?: number | null;
  apgar_5?: number | null;
  maternal_complications?: string | null;
  newborn_complications?: string | null;
  notes?: string | null;
}

export interface MaternityDeliveryResponse extends MaternityDeliveryUpsert {
  id: string;
  clinic_id: string;
  visit_id: string;
  recorded_by: string;
  recorded_at: string;
  record_status: 'DRAFT' | 'SIGNED' | 'AMENDED' | 'VOIDED';
  signed_at?: string | null;
  void_reason?: string | null;
}

export interface PostnatalNoteCreate {
  subject: 'MOTHER' | 'BABY';
  note: string;
}

export interface PostnatalNoteResponse extends PostnatalNoteCreate {
  id: string;
  clinic_id: string;
  visit_id: string;
  added_by: string;
  added_at: string;
}

export interface FamilyPlanningEventCreate {
  commodity: 'IMPLANT' | 'IUD' | 'INJECTABLE' | 'PILL' | 'CONDOM' | 'OTHER';
  notes?: string | null;
}

export interface FamilyPlanningEventResponse extends FamilyPlanningEventCreate {
  id: string;
  clinic_id: string;
  visit_id: string;
  added_by: string;
  added_at: string;
}

const purpose_of_use = PurposeOfUse.TREATMENT;

export const maternityService = {
  async getQueue(status?: VisitStatus): Promise<VisitResponse[]> {
    const params = status ? { status } : undefined;
    const response = await client.get('/v1/maternity/queue', { params });
    return response.data;
  },

  async getDelivery(visitId: string): Promise<MaternityDeliveryResponse | null> {
    const response = await client.get(`/v1/maternity/visits/${visitId}/delivery`, {
      params: {
        purpose_of_use,
        justification: 'Maternity delivery view',
      },
    });
    return response.data ?? null;
  },

  async upsertDelivery(
    visitId: string,
    payload: MaternityDeliveryUpsert
  ): Promise<MaternityDeliveryResponse> {
    const response = await client.post(
      `/v1/maternity/visits/${visitId}/delivery`,
      payload
    );
    return response.data;
  },

  async listPostnatalNotes(visitId: string): Promise<PostnatalNoteResponse[]> {
    const response = await client.get(
      `/v1/maternity/visits/${visitId}/postnatal-notes`,
      {
        params: {
          purpose_of_use,
          justification: 'Maternity postnatal notes',
        },
      }
    );
    return response.data;
  },

  async addPostnatalNote(
    visitId: string,
    payload: PostnatalNoteCreate
  ): Promise<PostnatalNoteResponse> {
    const response = await client.post(
      `/v1/maternity/visits/${visitId}/postnatal-notes`,
      payload
    );
    return response.data;
  },

  async listFamilyPlanningEvents(
    visitId: string
  ): Promise<FamilyPlanningEventResponse[]> {
    const response = await client.get(
      `/v1/maternity/visits/${visitId}/family-planning`,
      {
        params: {
          purpose_of_use,
          justification: 'Maternity family planning',
        },
      }
    );
    return response.data;
  },

  async addFamilyPlanningEvent(
    visitId: string,
    payload: FamilyPlanningEventCreate
  ): Promise<FamilyPlanningEventResponse> {
    const response = await client.post(
      `/v1/maternity/visits/${visitId}/family-planning`,
      payload
    );
    return response.data;
  },
};

File: app/services/maternity_service.py
# app/services/maternity_service.py
from datetime import datetime, timezone
from uuid import UUID

from fastapi import HTTPException

from app.models.maternity_delivery_record import MaternityDeliveryRecord
from app.models.maternity_postnatal_note import MaternityPostnatalNote
from app.models.family_planning_event import FamilyPlanningEvent
from app.models.pregnancy_episode import PregnancyEpisode
from app.models.visit import Visit
from app.services.follow_up_workflow_service import FollowUpWorkflowService
from app.shared.enums import RecordStatus, VisitServiceLine, VisitStatus


class MaternityService:
    def __init__(self, db):
        self.db = db

    def _get_visit_for_maternity(self, *, visit_id: UUID, clinic_id: UUID, actor_id: UUID) -> Visit:
        visit = (
            self.db.query(Visit)
            .filter(
                Visit.id == visit_id,
                Visit.clinic_id == clinic_id,
            )
            .first()
        )
        if not visit:
            raise HTTPException(status_code=404, detail="Visit not found")
        if visit.service_line != VisitServiceLine.MATERNITY:
            raise HTTPException(status_code=400, detail="Visit is not maternity")
        if visit.assigned_doctor_id != actor_id:
            raise HTTPException(status_code=403, detail="Visit not assigned to you")
        if visit.status in {VisitStatus.CANCELLED}:
            raise HTTPException(status_code=400, detail="Visit is cancelled")
        return visit

    def get_delivery(
        self, *, clinic_id: UUID, visit_id: UUID, actor_id: UUID
    ) -> MaternityDeliveryRecord | None:
        self._get_visit_for_maternity(visit_id=visit_id, clinic_id=clinic_id, actor_id=actor_id)
        return (
            self.db.query(MaternityDeliveryRecord)
            .filter(
                MaternityDeliveryRecord.clinic_id == clinic_id,
                MaternityDeliveryRecord.visit_id == visit_id,
            )
            .first()
        )

    def upsert_delivery(
        self,
        *,
        clinic_id: UUID,
        visit_id: UUID,
        actor_id: UUID,
        payload,
    ) -> MaternityDeliveryRecord:
        visit = self._get_visit_for_maternity(visit_id=visit_id, clinic_id=clinic_id, actor_id=actor_id)
        episode = None
        if payload.episode_id:
            episode = (
                self.db.query(PregnancyEpisode)
                .filter(
                    PregnancyEpisode.id == payload.episode_id,
                    PregnancyEpisode.clinic_id == clinic_id,
                )
                .first()
            )
            if not episode:
                raise HTTPException(status_code=404, detail="Episode not found")

        record = (
            self.db.query(MaternityDeliveryRecord)
            .filter(
                MaternityDeliveryRecord.clinic_id == clinic_id,
                MaternityDeliveryRecord.visit_id == visit_id,
            )
            .first()
        )

        now = datetime.now(timezone.utc)
        if not record:
            record = MaternityDeliveryRecord(
                clinic_id=clinic_id,
                visit_id=visit_id,
                episode_id=episode.id if episode else None,
                recorded_by=actor_id,
                recorded_at=now,
                record_status=RecordStatus.DRAFT,
            )
            self.db.add(record)

        if record.record_status == RecordStatus.SIGNED:
            raise HTTPException(status_code=409, detail="Delivery record already signed")

        record.episode_id = episode.id if episode else None
        record.delivered_at = payload.delivered_at
        record.mode_of_delivery = payload.mode_of_delivery
        record.outcome = payload.outcome
        record.baby_sex = payload.baby_sex
        record.baby_weight_kg = payload.baby_weight_kg
        record.apgar_1 = payload.apgar_1
        record.apgar_5 = payload.apgar_5
        record.maternal_complications = payload.maternal_complications
        record.newborn_complications = payload.newborn_complications
        record.notes = payload.notes

        if payload.action == "SIGN":
            record.record_status = RecordStatus.SIGNED
            record.signed_at = now
            FollowUpWorkflowService(self.db).complete_linked_follow_up_on_visit_sign(
                clinic_id=clinic_id,
                actor_id=actor_id,
                visit_id=visit.id,
                patient_id_canonical=visit.patient_id,
                linked_follow_up_id=visit.linked_follow_up_id,
                signed_at=now,
                commit=False,
            )
        else:
            record.record_status = RecordStatus.DRAFT

        self.db.add(record)
        self.db.commit()
        self.db.refresh(record)
        return record

    def list_postnatal_notes(
        self, *, clinic_id: UUID, visit_id: UUID, actor_id: UUID
    ) -> list[MaternityPostnatalNote]:
        self._get_visit_for_maternity(visit_id=visit_id, clinic_id=clinic_id, actor_id=actor_id)
        return (
            self.db.query(MaternityPostnatalNote)
            .filter(
                MaternityPostnatalNote.clinic_id == clinic_id,
                MaternityPostnatalNote.visit_id == visit_id,
            )
            .order_by(MaternityPostnatalNote.added_at.desc())
            .all()
        )

    def add_postnatal_note(
        self,
        *,
        clinic_id: UUID,
        visit_id: UUID,
        actor_id: UUID,
        payload,
    ) -> MaternityPostnatalNote:
        self._get_visit_for_maternity(visit_id=visit_id, clinic_id=clinic_id, actor_id=actor_id)
        note = MaternityPostnatalNote(
            clinic_id=clinic_id,
            visit_id=visit_id,
            subject=payload.subject,
            note=payload.note,
            added_by=actor_id,
            added_at=datetime.now(timezone.utc),
        )
        self.db.add(note)
        self.db.commit()
        self.db.refresh(note)
        return note

    def list_family_planning_events(
        self, *, clinic_id: UUID, visit_id: UUID, actor_id: UUID
    ) -> list[FamilyPlanningEvent]:
        self._get_visit_for_maternity(visit_id=visit_id, clinic_id=clinic_id, actor_id=actor_id)
        return (
            self.db.query(FamilyPlanningEvent)
            .filter(
                FamilyPlanningEvent.clinic_id == clinic_id,
                FamilyPlanningEvent.visit_id == visit_id,
            )
            .order_by(FamilyPlanningEvent.added_at.desc())
            .all()
        )

    def add_family_planning_event(
        self,
        *,
        clinic_id: UUID,
        visit_id: UUID,
        actor_id: UUID,
        payload,
    ) -> FamilyPlanningEvent:
        self._get_visit_for_maternity(visit_id=visit_id, clinic_id=clinic_id, actor_id=actor_id)
        event = FamilyPlanningEvent(
            clinic_id=clinic_id,
            visit_id=visit_id,
            commodity=payload.commodity,
            notes=payload.notes,
            added_by=actor_id,
            added_at=datetime.now(timezone.utc),
        )
        self.db.add(event)
        self.db.commit()
        self.db.refresh(event)
        return event
